### PR TITLE
Preserve inverse transform image scope

### LINKED ISSUE #1480: inverse transforms lose include/exclude scope and mutate unselected images
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Bug summary

Inverse transforms do not preserve the original `include` / `exclude` scope. If a transform is applied to only a subset of images, `apply_inverse_transform()` can apply the inverse to images that were never transformed, corrupting them. This affects intensity transforms such as `Gamma`, and also appears to affect spatial transforms such as `Flip`.

## Root cause

`AppliedTransform` currently records only the transform class name and sampled params:

```
AppliedTransform(name=type(self).__name__, params=params)
```

Then `get_inverse_transform()` reconstructs the inverse using only those params:

```
steps.append(instance.inverse(trace.params))
```

As a result, the original image filter is lost.

For example, `Gamma.inverse()` returns:

```
_GammaInverse(log_gamma=params["log_gamma"], copy=False)
```

with no `include` / `exclude`, so `_GammaInverse` applies to all scalar images.

## Suggested fix

Record the transform scope in history and reapply it to the rebuilt inverse transform.

Possible options:

Store the actual image names touched by the forward transform, and set `include` on the inverse to those names.

Regression tests should cover at least:

- `Gamma(log_gamma=..., include=["a"])`
- `Gamma(log_gamma=..., exclude=["b"])`
- a spatial transform such as `Flip(axes=(0,), include=["a"])`

### Code for reproduction

```python
import torch
import torchio as tio

a = torch.arange(8.).reshape(1, 2, 2, 2)
b = torch.arange(100, 108.).reshape(1, 2, 2, 2)

subject = tio.Subject(
    a=tio.ScalarImage(a.clone()),
    b=tio.ScalarImage(b.clone()),
)

out = tio.Gamma(log_gamma=0.5, include=["a"])(subject)
back = out.apply_inverse_transform()

print(torch.allclose(back.a.data, a))  # True
print(torch.allclose(back.b.data, b))  # False
print(out.applied_transforms)
```

### Actual outcome

The inverse transform is reconstructed without the original `include` / `exclude` information, so it defaults to applying to all eligible images.

### Error messages

```python-traceback
[AppliedTransform(name="Gamma", params={"log_gamma": 0.5})]
```

### Expected outcome

The inverse transform should only be applied to the same image scope as the forward transform.
Images not touched by the forward transform should remain unchanged after `apply_inverse_transform()`.

### System info

```shell

```

### PR DESCRIPTION
**Pull request by gaoflow**
## Summary
- record the image names affected by each applied transform
- restrict rebuilt inverse transforms to those names when they exist on the current data
- add Gamma include/exclude and Flip include regression coverage

Fixes #1480.

## Tests
- uv run --group test pytest tests/test_gamma.py tests/test_flip.py -q -k 'inverse_respects_include_scope or inverse_respects_exclude_scope'
- uv run --group test pytest tests/test_gamma.py tests/test_flip.py tests/test_inverse.py tests/test_transforms_base.py -q
- uv run --group test pytest tests/test_bias_field.py tests/test_normalize.py tests/test_standardize.py tests/test_remap_labels.py tests/test_one_hot.py tests/test_sequential_labels.py tests/test_transpose.py tests/test_crop.py tests/test_pad.py -q
- uv run --group test pytest -m 'not slow' -q
- uv run --group quality ruff check
- uv run --group quality ruff format --check
- uv run python -m compileall -q src tests
- uvx --with tox-uv tox -e lint,types,format

AI assistance was used under my direction.

### CODE REVIEW

File: src/torchio/transforms/inverse.py [hunk 1]
```diff
@@ -52,10 +55,27 @@ def get_inverse_transform(
...
-        steps.append(instance.inverse(trace.params))
+        inverse = instance.inverse(trace.params)
+        if trace.image_names is not None and current_image_names is not None:
+            matching_names = [
+                name for name in trace.image_names if name in current_image_names
+            ]
+            if matching_names:
+                inverse.include = matching_names
+                inverse.exclude = None
+        steps.append(inverse)
```
[the remark this answers is not in the document]
Comment (binggao1230) [reply src/torchio/transforms/inverse.py:58]: Handled in the latest push. `_restrict_inverse_to_trace_images()` now returns whether the inverse should be appended; when recorded image names have no match in the current data, that inverse step is skipped instead of falling back to all images. I also added a regression test for the removed-image case and adjusted the copied-history test to require matching image names.

File: src/torchio/transforms/inverse.py [hunk 2]
```diff
@@ -52,12 +55,43 @@ def get_inverse_transform(
...
+    current_image_names: set[str] | None,
+) -> bool:
+    if trace.image_names is None or current_image_names is None:
+        return True
+    matching_names = [name for name in trace.image_names if name in current_image_names]
+    if not matching_names:
+        return False
+    inverse.include = matching_names
+    inverse.exclude = None
+    return True
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/inverse.py]: [Generated by a coding agent]

---

Good catch, and I verified it empirically:

- On `main`, the documented TTA snippet (history copied onto `pred=...`, then `apply_inverse_transform(ignore_intensity=True)`) correctly inverts the spatial transform.
- On this branch the inverse is silently skipped: `restored.pred` equals the still-augmented data, because the forward recorded `image_names=['t1']` and there is no overlap with `{pred}`.

On the suggested remedy, though, I'd avoid the intensity/non-intensity split with a "fall back to all current images" branch: that reintroduces the over-application this PR is trying to prevent (for spatial transforms) and is hard to reason about.

A simpler fix that satisfies both your review comments at once: record the forward transform's own `include`/`exclude` (not the resolved image names) and set them on the inverse. Then:

- The original `Gamma(include=['a'])` corruption is fixed: the inverse gets `include=['a']`.
- TTA is preserved: a forward with no scope records `include=None`, so the inverse applies to whatever images are present, including a renamed `pred`.
- The removed-image case from the earlier thread stays safe without any explicit skip: `_get_images` only returns images that exist, so `include=['a']` on a subject without `a` is a natural no-op.
- No `data=` threading, no `current_image_names`, no intensity/spatial branching.

Intensity inverses re-apply their own scalar-only filter, so label maps still won't be touched.

File: src/torchio/transforms/inverse.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/inverse.py:58]: [Generated by a coding agent]

---

For the record: the skip-when-no-overlap behavior added in response to this comment is exactly what introduced the TTA regression flagged in the later review (history copied onto a differently-named `pred` image gets its spatial inverse silently dropped).

Recording the forward `include`/`exclude` instead avoids needing a skip at all. The removed-image case you raised here stays safe because `_get_images` only returns images that exist, so an `include` naming an absent image is already a no-op.

**Review by fepegar — CHANGES_REQUESTED** [review]
[Generated by a coding agent]

---

Thanks for tackling this, the underlying bug (#1480) is real and the new include/exclude regression tests are valuable.

My one blocking concern is a regression in the documented test-time-augmentation workflow, which I verified empirically:

| scenario | `main` | this branch |
| --- | --- | --- |
| copy history onto a renamed `pred` image, then `apply_inverse_transform` | inverse applied ✅ | inverse silently skipped ❌ |

The TTA guide (`docs/how-to/tta.md`, lines 37-46) wraps the model output as `pred=tio.ScalarImage(...)`, copies `applied_transforms`, and inverts. Because the forward `Flip` recorded `image_names=['t1']` and there's no overlap with `{pred}`, `_restrict_inverse_to_trace_images` returns `False` and the spatial inverse is dropped, so predictions are never mapped back to the original space.

Two things mask this in the PR:
- `tests/test_flip.py::test_inverse_on_image_via_subject` was changed to rename `pred` -> `t1` so the names match, which hides the regression rather than addressing it.
- `docs/how-to/tta.md` still documents the now-broken `pred=...` pattern and isn't updated.

### Suggested approach

Record the forward transform's own `include`/`exclude` in the trace and set them on the inverse, instead of recording resolved image names and intersecting them with the current data. This:

- fixes the original corruption (`Gamma(include=['a'])` -> inverse `include=['a']`);
- preserves TTA, since a scopeless forward records `include=None` and the inverse applies to whatever images are present (including a renamed `pred`);
- keeps the removed-image case safe with no explicit skip, because `_get_images` only returns images that exist, so an `include` naming an absent image is a no-op;
- drops the `data=` threading, `current_image_names`, `_get_image_names`, and the intensity/spatial special-casing entirely.

Intensity inverses re-apply their own scalar-only filter, so label maps stay untouched either way.

If you'd like, I'm happy to push a commit implementing this and add a TTA regression test (renamed prediction image) plus keep the original `pred`-named `test_inverse_on_image_via_subject`.

**Comment by fepegar** [comment]
Thanks for the PR, @gaoflow. Would you like to keep addressing the clankers' reviews? Otherwise I'm happy to take over.

**Comment by binggao1230** [comment]
I kept going and pushed `16a9175b` with the suggested approach:

- `AppliedTransform` now records the forward transform’s `include` / `exclude` scope instead of resolved image names.
- inverse reconstruction applies that recorded scope to the inverse transform and drops the `data=` / current-image-name intersection path.
- restored the copied-history prediction case to use a renamed `pred` image, so the documented TTA pattern is covered again.

Verification:

```bash
uv run tox -e test -- tests/test_flip.py tests/test_gamma.py tests/test_inverse.py tests/test_pad.py -q
uv run tox -e lint
uv run tox -e format
uv run tox -e prek
```

File: src/torchio/transforms/spatial/crop_or_pad.py [hunk 3]
```diff
@@ -512,6 +512,8 @@ def _apply_lazy_ops(
     ) -> None:
         """Apply lazy pad/crop and record history."""
         images = _get_images(subject, self.include, self.exclude)
+        include = list(self.include) if self.include is not None else None
+        exclude = list(self.exclude) if self.exclude is not None else None
```
Comment (fepegar) [inline src/torchio/transforms/spatial/crop_or_pad.py]:
```suggestion
        include = None if self.include is None else list(self.include)
        exclude = None if self.exclude is None else list(self.exclude)
```

**Review by fepegar — DISMISSED** [review]
Looking good. Thanks for your contribution, @gaoflow. Thanks also for disclosing your usage of AI.

**Comment by fepegar** [comment]
@all-contributors please add @gaoflow for bug, code.
